### PR TITLE
Fuse 'Add' operator into FP16 Conv

### PR DESCRIPTION
**Pull request by chenfucn**
### Description
Adding 'Add' functionality to FP16 Conv operator. It takes a tensor that has the same shape of the output tensor, and add it to the result tensor.

### Motivation and Context
Needed to run Resnet 50

### CODE REVIEW

**Comment by chenfucn** [comment]
Two commits now, first one 34b1b99b2d34fc72644a02b0647f91e83f057243 is just https://github.com/microsoft/onnxruntime/pull/15245 or else arm64 compilation fail. So ignore that when you review.

File: onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc [hunk 1]
```diff
@@ -84,6 +100,8 @@ void OpSet_Internal_NHWC_ONNX::ForEachSchema(const std::function<void(ONNX_NAMES
   REGISTER_NHWC_SCHEMA(fn, QLinearConv, 10);
   REGISTER_NHWC_SCHEMA_FROM_MSDOMAIN(fn, QLinearAveragePool, 1);
   REGISTER_NHWC_SCHEMA_FROM_MSDOMAIN(fn, QLinearConvTranspose, 1);
+  RegisterNHWCSchemaWithSumActivation(
+      fn, contrib::GetOpSchema<contrib::ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Microsoft, 1, FusedConv)>());
```
Comment (skottmckay) [inline onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc:104]: The existing operators here are 1:1 with ONNX operators and their opsets, with the only difference being the conversion of NCHW inputs/outputs to NHWC. This relationship is leveraged by layout transformation as we can convert the ONNX layout sensitive operator to the NHWC version in the internal domain. FusedConv doesn't seem to quite fit that if it's not mirroring an ONNX operator. 

If we want to start using the kMSInternalNHWCDomain domain for NHWC contrib ops as well we'd need to make sure that that doesn't invalidate any assumptions made in other code (e.g. graph partitioning, transpose optimizer) about the ops in that domain. Mixing in ONNX and contrib ops potentially makes the domain a little more confusing though.

There are some existing contrib ops that support NHWC layout which are used by nhwc_transformer.cc/h. They have a 'channels_last' attribute. Would it be more consistent to take that approach so the existing NCHW FusedConv and new NCHW FusedConv simply differ by the usage of that attribute? That way they're in the same domain and like all contrib ops have a fixed opset of 1.

File: onnxruntime/contrib_ops/cpu/cpu_contrib_kernels.cc [hunk 2]
```diff
@@ -17,7 +17,7 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, float, ExpandDims);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, float, FusedConv);
 #ifdef MLAS_F16VEC_INTRINSICS_SUPPORTED
-class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, MLFloat16, FusedConv);
+class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSInternalNHWCDomain, 1, MLFloat16, FusedConv);
```
Comment (skottmckay) [inline onnxruntime/contrib_ops/cpu/cpu_contrib_kernels.cc:20]: Won't changing the domain break any existing models using this operator?

Comment (chenfucn) [reply onnxruntime/contrib_ops/cpu/cpu_contrib_kernels.cc:20]: there is no existing code utilizing this operator yet. But I get what you mean. I would have to create a new schema I guess?

File: onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc [hunk 3]
```diff
@@ -383,5 +383,41 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
...
     OpSchema().FillUsing(ConvOpSchemaGenerator()));
+
+ONNX_MS_OPERATOR_SET_SCHEMA(NhwcFusedConv, 1,
+                            OpSchema()
+                                .SetDoc(R"DOC(
+NhwcFusedConv is a float16 Conv operator with optional activation and add operators fused in.
+It also processes both NHWC and NCHW format.
+)DOC")
+                                .Attr("channels_last", "0(default) when input tensor is NCHW, else the input is NHWC", AttributeProto::INT, static_cast<int64_t>(0))
+                                .Attr("auto_pad", "", AttributeProto::STRING, std::string("NOTSET"))
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc:394]: The name already has Nhwc, which indicates the op is channel last.

File: onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc [hunk 4]
```diff
@@ -383,5 +383,41 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
...
+                                .Input(0, "X", "", "T")
+                                .Input(1, "W", "", "T")
+                                .Input(2, "B", "", "T", OpSchema::Optional)
+                                .Input(3, "Z", "Tensor to be added to the output, must be the same shape and format as the output tensor.", "T", OpSchema::Optional)
+                                .Output(0, "Y", "", "T")
+                                .TypeConstraint("T", {"tensor(float16)"}, "Constrain input and output types to float tensors")
+                                //.TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+                                //  ONNX_NAMESPACE::propagateElemTypeFromInputToOutput(ctx, 0, 0);
+                                //  ONNX_NAMESPACE::convPoolShapeInference(ctx, true, false, 0, 1);
+                                //}));
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc:411]: nit: remove the commented out code.

File: onnxruntime/core/providers/cpu/fp16/fp16_conv.cc [hunk 5]
```diff
@@ -239,7 +241,7 @@ Status FusedConvFp16::Compute(OpKernelContext* context) const {
   // TODO!!
   // This tensor should be added to the result before activation is applied
   // We need to augment the post processor to accept an addition operation.
-  // const Tensor* Sum = num_inputs >= 4 ? context->Input<Tensor>(3) : nullptr;
+  const Tensor* Sum = num_inputs >= 4 ? context->Input<Tensor>(3) : nullptr;
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:244]: nit: remove the TODO

File: onnxruntime/core/providers/cpu/fp16/fp16_conv.cc [hunk 6]
```diff
@@ -532,7 +545,8 @@ Status FusedConvFp16::Compute(OpKernelContext* context) const {
           }
 
           // TODO!! add Sum tensor to activation
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:547]: >// TODO!! add Sum tensor to activation [](http://example.com/codeflow?start=10&length=38)

nit: remove to do

File: onnxruntime/core/providers/cpu/fp16/fp16_conv.cc [hunk 7]
```diff
@@ -34,9 +34,9 @@ using ConvPadVector = ConvAttributes::ConvPadVector;
  *
  * Add is performed BEFORE activation.
  *
- * The implementation runs faster with NHWC. By default, it converts NCHW to NHWC
- * before processing, and convert the result back. It can take NHWC tensors directly.
- * Use operator attribute 'channels_last' to specify that the data layout is NHWC.
+ * The implementation runs faster with NHWC. In base class, where channel_last_ is
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/fp16/fp16_conv.cc:37]: >In base class [](http://example.com/codeflow?start=45&length=13)

what does base class refer to?

File: onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc:394]: Which activations are you wanting to support? The registration for the NHWC Conv in the internal NHWC domain has both activation and activation_params. 

[https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc#L90](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc#L90)

[https://github.com/microsoft/onnxruntime/blob/198994d01d4e226fd3a8816e8b2fad00c84a691c/onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc#L42-L43](https://github.com/microsoft/onnxruntime/blob/198994d01d4e226fd3a8816e8b2fad00c84a691c/onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc#L42-L43)

So if you only need to add an NHWC format Conv that can fuse with an activation, the existing registration gives you one based on the ONNX schema (which includes float16) with ONNX type/shape inferencing.

Comment (chenfucn) [reply onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc:394]: Two things. 1, FusedConv have not only activation, but also an Add operator fused in.  2. I don't fully understand the rationale of this part to be honest. In my mind the schema is a contract. And our implementations of the operators must adhere to the contracts. It seems to me only the xnnpack EP's internal nhwc conv operator support activation attributes? What about other implementations?

**Review by yufenglee — APPROVED** [review]

**Pull request merged**